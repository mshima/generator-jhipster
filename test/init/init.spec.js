const path = require('path');
const expect = require('expect');
const { access } = require('fs/promises');
const normalizePath = require('normalize-path');

const { testBlueprintSupport } = require('../support');
const { skipPrettierHelpers: helpers } = require('../utils/utils');
const { defaultConfig } = require('../../generators/init/config');

const initGeneratorPath = path.join(__dirname, '../../generators/init');

const testDefaultConfg = { ...defaultConfig, jhipsterVersion: '0.0.0' };

describe('JHipster init generator', () => {
  describe('with default options', () => {
    let runResult;
    before(async () => {
      runResult = await helpers.run(initGeneratorPath);
    });
    it('should create expected files', () => {
      expect(runResult.getStateSnapshot()).toMatchSnapshot();
    });
  });
  describe('with defaults option', () => {
    let runResult;
    before(async () => {
      runResult = await helpers.run(initGeneratorPath).withOptions({ defaults: true });
    });
    it('should load default config into the generator', () => {
      expect(runResult.generator).toMatchObject(testDefaultConfg);
    });
  });
  describe('with custom prompt values', () => {
    let runResult;
    const promptValues = {
      prettierDefaultIndent: 4,
      prettierJavaIndent: 2,
    };
    describe('and default options', () => {
      before(async () => {
        runResult = await helpers.run(initGeneratorPath).withOptions({ baseName: 'jhipster' }).withPrompts(promptValues);
      });
      it('should write custom config to .yo-rc.json', () => {
        const yoFile = normalizePath(path.join(runResult.cwd, '.yo-rc.json'));
        expect(runResult.getSnapshot(file => file.path === yoFile)).toMatchSnapshot();
      });
    });
    describe('and defaults option', () => {
      before(async () => {
        runResult = await helpers.run(initGeneratorPath).withOptions({ baseName: 'jhipster', defaults: true }).withPrompts(promptValues);
      });
      it('should not write custom config to .yo-rc.json', () => {
        const yoFile = normalizePath(path.join(runResult.cwd, '.yo-rc.json'));
        expect(runResult.getSnapshot(file => file.path === yoFile)).toMatchSnapshot();
      });
    });
    describe('and skipPrompts option', () => {
      let runResult;
      before(async () => {
        runResult = await helpers.run(initGeneratorPath).withOptions({ skipPrompts: true, baseName: 'jhipster' }).withPrompts(promptValues);
      });
      it('should write default values to .yo-rc.json', () => {
        const yoFile = normalizePath(path.join(runResult.cwd, '.yo-rc.json'));
        expect(runResult.getSnapshot(file => file.path === yoFile)).toMatchSnapshot();
      });
    });
    describe('and existing config', () => {
      let runResult;
      before(async () => {
        runResult = await helpers
          .run(initGeneratorPath)
          .withOptions({ localConfig: { baseName: 'existing' } })
          .withPrompts(promptValues);
      });
      it('should not write custom prompt values', () => {
        const yoFile = normalizePath(path.join(runResult.cwd, '.yo-rc.json'));
        expect(runResult.getSnapshot(file => file.path === yoFile)).toMatchSnapshot();
      });
    });
    describe('and askAnswered option on an existing project', () => {
      let runResult;
      before(async () => {
        runResult = await helpers
          .run(initGeneratorPath)
          .withOptions({ askAnswered: true, localConfig: { baseName: 'existing' } })
          .withPrompts(promptValues);
      });
      it('should write custom prompt values', () => {
        const yoFile = normalizePath(path.join(runResult.cwd, '.yo-rc.json'));
        expect(runResult.getSnapshot(file => file.path === yoFile)).toMatchSnapshot();
      });
    });
  });
  describe('with custom', () => {
    describe('skipCommitHook option', () => {
      let runResult;
      before(async () => {
        runResult = await helpers.run(initGeneratorPath).withOptions({ skipCommitHook: true });
      });
      it('should create expected files', () => {
        expect(runResult.getStateSnapshot()).toMatchSnapshot();
      });
    });
  });
  describe('git feature', () => {
    describe('with default option', () => {
      let runResult;
      before(async () => {
        runResult = await helpers.run(initGeneratorPath);
      });
      it('should create .git', async () => {
        await expect(access(path.resolve(runResult.cwd, '.git'))).resolves.toBeUndefined();
      });
      it('should create 1 commit', async () => {
        const git = runResult.generator._createGit();
        await expect(git.log()).resolves.toMatchObject({
          total: 1,
          latest: { message: expect.stringMatching(/^Initial version of/) },
        });
      });
    });
    describe('with skipGit option', () => {
      let runResult;
      before(async () => {
        runResult = await helpers.run(initGeneratorPath).withOptions({ skipGit: true });
      });
      it('should not create .git', async () => {
        await expect(access(path.resolve(runResult.cwd, '.git'))).rejects.toMatchObject({ code: 'ENOENT' });
      });
    });
    describe('regenerating', () => {
      let runResult;
      before(async () => {
        runResult = await helpers.run(initGeneratorPath);
        runResult = await runResult.create(initGeneratorPath).withOptions({ skipPrettier: true, jhipsterVersion: '1.0.0' }).run();
      });
      it('should have 1 commit', async () => {
        const git = runResult.generator._createGit();
        await expect(git.log()).resolves.toMatchObject({ total: 1 });
      });
      it('should have uncommited files', async () => {
        const git = runResult.generator._createGit();
        await expect(git.diff()).resolves.toMatch(/\+ {4}"jhipsterVersion": "1\.0\.0"/);
      });
    });
  });
  describe('blueprint support', () => testBlueprintSupport('init'));
});
