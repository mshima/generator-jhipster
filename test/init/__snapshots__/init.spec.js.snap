// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JHipster init generator with custom prompt values and askAnswered option on an existing project should write custom prompt values 1`] = `
Object {
  ".yo-rc.json": Object {
    "contents": "{
  \\"generator-jhipster\\": {
    \\"baseName\\": \\"existing\\",
    \\"projectName\\": \\"JHipster project\\",
    \\"prettierDefaultIndent\\": 4,
    \\"prettierJavaIndent\\": 2,
    \\"jhipsterVersion\\": \\"0.0.0\\"
  }
}
",
    "stateCleared": "modified",
  },
}
`;

exports[`JHipster init generator with custom prompt values and default options should write custom config to .yo-rc.json 1`] = `
Object {
  ".yo-rc.json": Object {
    "contents": "{
  \\"generator-jhipster\\": {
    \\"baseName\\": \\"jhipster\\",
    \\"projectName\\": \\"JHipster project\\",
    \\"prettierDefaultIndent\\": 4,
    \\"prettierJavaIndent\\": 2,
    \\"jhipsterVersion\\": \\"0.0.0\\"
  }
}
",
    "stateCleared": "modified",
  },
}
`;

exports[`JHipster init generator with custom prompt values and defaults option should not write custom config to .yo-rc.json 1`] = `
Object {
  ".yo-rc.json": Object {
    "contents": "{
  \\"generator-jhipster\\": {
    \\"baseName\\": \\"jhipster\\",
    \\"jhipsterVersion\\": \\"0.0.0\\",
    \\"projectName\\": \\"JHipster project\\"
  }
}
",
    "stateCleared": "modified",
  },
}
`;

exports[`JHipster init generator with custom prompt values and existing config should not write custom prompt values 1`] = `
Object {
  ".yo-rc.json": Object {
    "contents": "{
  \\"generator-jhipster\\": {
    \\"baseName\\": \\"existing\\",
    \\"jhipsterVersion\\": \\"0.0.0\\",
    \\"projectName\\": \\"JHipster project\\"
  }
}
",
    "stateCleared": "modified",
  },
}
`;

exports[`JHipster init generator with custom prompt values and skipPrompts option should write default values to .yo-rc.json 1`] = `
Object {
  ".yo-rc.json": Object {
    "contents": "{
  \\"generator-jhipster\\": {
    \\"baseName\\": \\"jhipster\\",
    \\"jhipsterVersion\\": \\"0.0.0\\",
    \\"projectName\\": \\"JHipster project\\"
  }
}
",
    "stateCleared": "modified",
  },
}
`;

exports[`JHipster init generator with custom skipCommitHook option should create expected files 1`] = `
Object {
  ".editorconfig": Object {
    "stateCleared": "modified",
  },
  ".gitattributes": Object {
    "stateCleared": "modified",
  },
  ".gitignore": Object {
    "stateCleared": "modified",
  },
  ".prettierignore": Object {
    "stateCleared": "modified",
  },
  ".prettierrc": Object {
    "stateCleared": "modified",
  },
  ".yo-rc.json": Object {
    "stateCleared": "modified",
  },
  "README.md": Object {
    "stateCleared": "modified",
  },
  "package.json": Object {
    "stateCleared": "modified",
  },
}
`;

exports[`JHipster init generator with default options should create expected files 1`] = `
Object {
  ".editorconfig": Object {
    "stateCleared": "modified",
  },
  ".gitattributes": Object {
    "stateCleared": "modified",
  },
  ".gitignore": Object {
    "stateCleared": "modified",
  },
  ".huskyrc": Object {
    "stateCleared": "modified",
  },
  ".lintstagedrc.js": Object {
    "stateCleared": "modified",
  },
  ".prettierignore": Object {
    "stateCleared": "modified",
  },
  ".prettierrc": Object {
    "stateCleared": "modified",
  },
  ".yo-rc.json": Object {
    "stateCleared": "modified",
  },
  "README.md": Object {
    "stateCleared": "modified",
  },
  "package.json": Object {
    "stateCleared": "modified",
  },
}
`;
